### PR TITLE
[Daniel Ogbuti]-Sage/Effective Smart Contract Testing using Chai

### PR DESCRIPTION
**Pull request by dahnny**
### Thanks for submitting this tutorial for Celo Sage

Before you hit "Create Pull Request", make sure you have read the [Tutorial Contribution Guidelines](https://docs.celo.org/community/celo-sage/contribution-guidelines) and reviewed the [Sample Tutorial Structure](https://docs.celo.org/community/celo-sage/tutorial-structure).

Please confirm the following before submitting your tutorial for review -

- [x] My tutorial follows the [Tutorial Contribution Guidelines](https://docs.celo.org/community/celo-sage/contribution-guidelines).
- [x] The title of the Pull Request matches the title of my tutorial.
- [x] I've followed the tutorial template structure.
- [x] I've proofread the tutorial for spelling and grammar mistakes.
- [x] I've thoroughly explained the code, with inline code comments and text explanations.
- [x] I've run a linter & prettifier to format my code blocks.
- [x] I've run the tutorial to test it and the code works.

> NOTE: We cannot review your Pull Request if it does not meet the standard outlined above.
> Reviewers - @joenyzio & @viral-sangani

### CODE REVIEW

File: blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+It's worth noting that the use of `const` indicates that these variables are declared as constants and cannot be reassigned once they are initialized. This helps ensure the stability and consistency of the testing environment.
+
+Next, you are going to define two test suites and set a longer timeout for the test cases in the `MyNFT` suite. This allows the test cases in the `MyNFT` suite to take longer to execute, which may be necessary for more complex test cases or for cases where the blockchain network is slow to respond. 
+
+```js
+describe("MyNftMarket", function () {});
+
+describe("MyNFT", function () {
+  this.timeout(50000);
+```
```
Comment (yafiabiyyu) [inline blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md:240]:
```suggestion
```js
describe("MyNftMarket", function () {});

describe("MyNFT", function(){
    this.timeout(50000)
    // The next code is below
})
```
```

File: blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+ let myNFT;
+  let owner;
+  let acc1;
+  let acc2;
+  let market;
+
+  it("Should create and execute market sales", async function () {
+    const MyNFT = await ethers.getContractFactory("MyNFT");
+    [owner, acc1, acc2] = await ethers.getSigners();
+```
```
Comment (yafiabiyyu) [inline blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md:264]:
```suggestion
```js
let myNFT;
let owner;
let acc1;
let acc2;
let market;

it("Should create and execute market sales", async function(){
    const MyNFT = await ethers.getContractFactory("MyNFT");
    [owner, acc1, acc2] = await ethers.getSigners();
    // The next code is below
})
```
```

File: blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md (hunk 1 above)
Comment (yafiabiyyu) [inline blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md:240]:
```suggestion
```js
describe("MyNftMarket", function () {});

describe("MyNFT", function(){
    this.timeout(50000)
});
```
```

File: blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+Next, we are going to create new instances of two smart contracts, `MyNFT` and `MyNftMarket`, and ensure that they are deployed to the blockchain before any tests are run.
+
+```js
+    const MyNftMarket = await ethers.getContractFactory("MyNftMarket");
+    market = await MyNftMarket.deploy();
+    await market.deployed();
+
+    myNFT = await MyNFT.deploy(market.address);
+    await myNFT.deployed();
+```
```
Comment (yafiabiyyu) [inline blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md:287]:
```suggestion
```js
//Previous code
const MyNftMarket = await ethers.getContractFactory("MyNftMarket");
market = await MyNftMarket.deploy();
await market.deployed();

myNFT = await MyNFT.deploy(market.address);
await myNFT.deployed();
```
```

File: blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+You then wait for the deployment of the MyNFT contract to complete by calling the `deployed` function on the `MyNFT` instance.
+
+Following this, you would write a test that will create and mint a new NFT to the account that called the `mint` function, with a metadata URL of "https://example.com/1" and an auction price of 1 Ether.
+
+```js
+    const auctionPrice = ethers.utils.parseUnits("1", "ether");
+
+    const nftContractAddress = myNFT.address;
+    const transaction = await myNFT.mint("https://example.com/1");
+```
```
Comment (yafiabiyyu) [inline blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md:304]:
```suggestion
```js
// Previous code
const auctionPrice = ethers.utils.parseUnits("1", "ether");

const nftContractAddress = myNFT.address;
const transaction = await myNFT.mint("https://example.com/1");
```
```

File: blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+
+The `mint` function is then called on the `myNFT` contract instance with the argument "https://example.com/1". This function call creates a new NFT with the given metadata URL and mints it to the account that called the function. The `await` keyword is used to wait for the transaction to be mined on the blockchain and a transaction object is returned, which contains information about the transaction such as the transaction hash and gas used.
+
+After creating a new NFT using the `mint` function in the previous code block, this code is used to list the newly created NFT for sale on the `MyNftMarket` contract instance.
+
+```js
+    await market.listToken(nftContractAddress, 1, auctionPrice);
+    const tx = await transaction.wait();
+    const tokenId =  tx.events[0].args[2].toNumber();
+```
```
Comment (yafiabiyyu) [inline blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md:318]:
```suggestion
```js
await market.listToken(nftContractAddress, 1, auctionPrice);
const tx = await transaction.wait();
const tokenId =  tx.events[0].args[2].toNumber();
```
```

File: blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+
+The `tokenId` variable is then assigned the ID of the newly listed NFT by accessing the `args` object of the first event emitted by the `wait` function call. The `toNumber()` function is used to convert the ID from a BigNumber object to a regular number.
+
+Following this, this code is used to simulate two different buyers purchasing the NFT that was listed for sale in the previous code block.
+
+```js
+    await market.connect(acc1).buyToken(0, { value: auctionPrice });
+    const t = await myNFT.connect(acc1).resaleApproval(tokenId);
+    await market.connect(acc2).buyToken(0, { value: auctionPrice });
+```
```
Comment (yafiabiyyu) [inline blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md:332]:
```suggestion
```js
// Previous code
await market.connect(acc1).buyToken(0, { value: auctionPrice });
const t = await myNFT.connect(acc1).resaleApproval(tokenId);
await market.connect(acc2).buyToken(0, { value: auctionPrice });
```
```

File: blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+    for (let index = 0; index < listingLength; index++) {
+      const listing = await market.getListing(index);
+      const [seller, owner, token, tokenId, price, sold] = listing;
+      // console.log(seller, owner, token, tokenId);
+      const tokenUri = await myNFT.tokenURI(tokenId);
+      console.log(tokenUri);
+        }
+    });
+
+```
```
Comment (yafiabiyyu) [inline blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md:354]:
```suggestion
```js
// Previous code
const listingLength = await market.getListingLength();

for(let index = 0; index < listingLength; index++) {
    const listing = await market.getListing(index);
    const [seller, owner, token, tokenId, price, sold] = listing;
    const tokenUri = await myNFT.tokenURI(tokenId);
    console.log(tokenURI);
}
```
```

**Review by Ishanpathak1 — APPROVED** [review]
Approved @developerkunal

**Review by Ishanpathak1 — APPROVED** [review]
Approved @developerkunal

**Pull request merged**